Docs/DOC-1068 Schema updates (#912)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/spatialos-concepts/schema.md

<%(TOC)%>
# Schema

Schema is a set of definitions which represent your game's classes in SpatialOS. SpatialOS uses schema to generate APIs specific to the components in your project. You can then use these APIs in your game's [worker types]({{urlRoot}}//content/glossary#spatialos-component) so their instances can interact with [SpatialOS entity components]({{urlRoot}}/content/glossary#spatialos-component). </br>

Schema is defined in `.schema` files and written in schemalang.  When you use the GDK, the schema files and their contents are generated and deleted automatically so you do not have to write or edit schema files manually. The GDK generates and deletes schema for you, when you start schema generation.

#### How to generate schema

There are two ways to generate schema for your project:

* **Generate schema (Full Scan)** 

>     To generate schema for all classes* in your project:<br/>

Reviewer: ```suggestion
    To generate schema for all classes in your project:<br/>
```

    In the Unreal Editor, on the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons), open the **Schema** drop-down menu and select **Schema (Full Scan)**.<br/> You must select **Schema (Full Scan)** the first time you generate schema for a project. 
    <br/> ![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button-full-scan.png)<br/>
    _Image: In the GDK toolbar in the Unreal Editor, select **Schema (Full Scan)**_<br/>
    <br/>When you select **Schema (Full Scan)**, the GDK automatically iterates through all classes* in your project to generate the schema files, and then updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase). <br/>

>     Run a full scan the first time you generate schema for your project, and whenever you need to generate schema for classes* that are not currently loaded by the Editor.<br/>For example: You need to select **Schema (Full Scan)** if you didn't generate schema after adding a Blueprint to a level, and that level is no longer open in the Editor.<br/><br/>

Reviewer: ```suggestion
    Run a full scan the first time you generate schema for your project, and whenever you need to generate schema for classes* that are not currently loaded by the Editor.<br/>For example: You need to select **Schema (Full Scan)** if you didn't generate schema after adding a new Blueprint to your game, and that Blueprint is no longer open in the Editor.<br/><br/>
```

* **Generate Schema (Iterative)**

    To generate schema for classes* that are currently loaded by the Editor: <br/>

>     Select  **Schema** in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through classes* that are currently loaded by the Editor, generates the schema files and updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase).<br/>

Reviewer: ```suggestion
    Select **Schema** in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through classes* that are currently loaded by the Editor, generates the schema files and updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase).<br/>
```

    ![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button.png)<br/>
    _Image: In the GDK toolbar in the Unreal Editor, select **Schema**_<br/><br/>

>     You must run an iterative scan when you add or change any classes* that you want to deploy to SpatialOS.

Reviewer: ```suggestion
    You must run an iterative scan when you add or change replicated properties in any classes* that you want to deploy to SpatialOS.
```

Reviewer: Do we need this sentence? Perhaps just refer to the *When to generate schema* section

Author: Nah, probably not. Good shout.

As the GDK automatically generates all the schema you need, you do not have to write or edit schema manually when using the GDK.

> \* Whenever you generate schema, the GDK automatically creates schema for classes with [replicated properties (Unreal documentation)](https://docs.unrealengine.com/en-US/Gameplay/Networking/Actors/Properties) or [RPCs (Unreal documentation)](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/RPCs). If a class does not have replicated properties or RPCs, then the GDK does not generate schema for it. 

#### When to generate schema

Select **Schema (Full Scan)** if you have: 

* Not generated any schema for your project.

> * Added a class with replicated properties to a level that is not currently loaded by the Editor.

Reviewer: ```suggestion
* Added a class with replicated properties that is not currently loaded by the Editor.
```

* Edited a class with replicated properties that is not currently loaded by the Editor.

Select **Schema** if you have:

> * Added a class with replicated properties to a level that is currently loaded by the Editor.

Reviewer: ```suggestion
* Added a class with replicated properties that is currently loaded by the Editor.
```

#### How to generate schema

There are two ways to generate schema for your project:

* **Generate schema (Full Scan)** 

>     To generate schema for all objects with [replicated properties (Unreal documentation)](https://docs.unrealengine.com/en-US/Gameplay/Networking/Actors/Properties) in your project, in the Unreal Editor, on the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons), open the **Schema** drop-down menu and select **Schema (Full Scan)**.<br/> You must select **Schema (Full Scan)** the first time you generate schema for a project. 

Reviewer: We also generate schema for objects that have RPCs. Here's unreal documentation on them - https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/RPCs

Reviewer: It's very much a nit, as 99% of the time any class that has an RPC will have replicated properties. You also don't need to regenerate schema when modifying RPCs, we just use the existence of RPCs to ensure the core data schema component is generated in the first place.

<br/><br/>

* **Generate Schema (Iterative)**

    To generate schema for objects with replicated properties that are currently loaded by the Editor, select  **Schema** in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through objects with replicated properties that are currently loaded by the Editor, generates the schema files and updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase).<br/>
    ![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button.png)<br/>
    _Image: In the GDK toolbar in the Unreal Editor, select **Schema**_<br/><br/>

    You must run an interative scan when you add or change any replicated properties that you want to deploy to SpatialOS.

As the GDK automatically generates all the schema you need, you do not have to write or edit schema manually when using the GDK.

#### When to generate schema

Select **Schema (Full Scan)** if you have: 

* Not generated any schema for your project.

> * Added an object with replicated properties to a level that is not currently loaded by the Editor.

Reviewer: References to `object` should probably be replaced with `class`.

#### How to generate schema

**Generating Schema for the first time**

> In the Unreal Editor, on the GDK toolbar, open the **Schema** drop-down menu and select **Schema (Full Scan)**. <br/>You must run a full scan the first time you generate Schema for a project. When you select Full Scan, the GDK automatically iterates through all classes with replicated properties to generate the required schema files and then updates the [SchemaDatabase](http://localhost:8080/reference/1.0/content/glossary#schemadatabase). <br/>

Reviewer 2: Nits: 
Do we need to say "required schema files" it reads a bit awkwardly?
My pedantry wants to make it "when you select **Schema (Full Scan)**".

```suggestion
In the Unreal Editor, on the GDK toolbar, open the **Schema** drop-down menu and select **Schema (Full Scan)**. <br/>You must run a full scan the first time you generate Schema for a project. When you select **Schema (Full Scan)**, the GDK automatically iterates through all classes with replicated properties to generate the schema files and then updates the [SchemaDatabase](http://localhost:8080/reference/1.0/content/glossary#schemadatabase). <br/>
```

![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button-full-scan.png)

_Image: In the GDK toolbar in the Unreal Editor, select **Schema (Full Scan)**_

**Generating Schema after running a full scan**

> To generate schema after you have run a full scan, select  **Schema** in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through classes with replicated properties that aren't already referenced in the SchemaDatabase. It then automatically generates the schema files and updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase).

Reviewer: This isn't quite true. The iterative scan analyses all objects currently in memory (currently loaded by the editor), and adds/updates schema definitions where necessary. If the asset isn't in memory though, it won't be added to the SchemaDatabase.

Reviewer: The *When to generate schema* section could be updated also.

> To generate schema after you have run a full scan, select  **Schema** in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through classes with replicated properties that aren't already referenced in the SchemaDatabase. It then automatically generates the required schema files and updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase).

Reviewer 2: ```suggestion
To generate schema after you have run a full scan, select  **Schema** in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through classes with replicated properties that aren't already referenced in the SchemaDatabase. It then automatically generates the schema files and updates the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase).
```
